### DISCUSSION TITLE
Callback function for onReject and logging for statistics

### DISCUSSION THREAD

**Discussion by lparikka** [orestbida/cookieconsent#90] [Ideas]
Hi
Thank you for this nice cookie consent plugin. 

I would like to get some statistics on the rejection of the cookies. As far as I understand, I can't get any feedback on rejection of the cookies as there is no onReject callback function. And it would be nice to be able to log the settings the users have selected for statistical reasons. 

Regards,
Lauri

**Comment by orestbida** [comment]
Hi @lparikka,

the "reject" action is treated as a special case of the accept action ("accept none", or "accept none except categories marked as necessary"). You can check #46 to find which categories have been accepted using the `onAccept`.

**Reply by lparikka** [reply]
Thank you for that information on handling reject. Is the onAccept run on every page load?

What do you think about my proposal to have a logging enhancement for getting the statistics of rejects and acceptances?

Lauri

**Reply by orestbida** [reply]
Yes, `onAccept` runs after each page load.

What do you need exactly? Like what data and what do you want to do with it.

**Reply by lparikka** [reply]
I would like to get statistics on which percentage of the visitors rejects the cookies to get some idea on how much of the visitors and page loads is missing from Google Analytics reports.

**Reply by orestbida** [reply]
There are multiple ways of how a user can reject or accept the coookie consent. 

3 types of “accept” levels.
- accept necessary only/reject all
- accept all
- accept custom selection

accept necessary only/reject all:
- user clicked the “reject” button inside the consent modal
- user clicked the “reject” button inside the settings modal
- user opened the settings modal, manually toggled off /on specific categories and clicked “save preferences”

accept all:
- user clicked the “accept” button inside the consent modal
- user clicked the “accept” button inside the settings modal
- user opened the settings modal, manually toggled on all categories and clicked “save preferences”

accept custom selection:
- user opened the settings modal, manually toggled off/on specific categories and clicked “save preferences”

So my questions are:
- When do you want to log preferences?
  - only the first time (cookie doesn't exist yet)
  - whenever they change preferences (you would need to determine the user by their ip/geolocation data)
  - when revision changes (if you plan on using the revision option)
- Perhaps you want something really basic like: only detect if user clicked either “accept” or “reject” button inside the consent modal ?

**Reply by lparikka** [reply]
At least simple like this:
- only the first time (cookie doesn't exist yet)
- user clicked Accept all or Reject all or Accept partly

**Reply by orestbida** [reply]
Ok. I could add a new (optional) parameter inside the `onAccept` method: `analytics_summary`
```javascript
onAccept : function(cookie_value, analytics_summary){ ... }
```

`analytics_summary` (an object) could have the following structure:
```javascript
{
  accept_type: 'accept_all',  //'accept_all' or 'accept_necessary' (reject all) or 'accept_selection'
  accepted_categories: ['necessary', 'analytics'],
  rejected_categories: ['advertisement']
}
```

We also need to make sure to only send this data **once**! We can use `.validCookie()` to check if the cookie exists or not:

```javascript
// if cookie doesn't exist => firstTime = true
const firstTime = !cc.validCookie('cc_cookie');
```

So if we put together the pieces, it would become something like this:
```javascript
const cc = initCookieConsent();
const firstTime = !cc.validCookie('cc_cookie');

cc.run({
    ..,
    onAccept: function(cookie_value, analytics_summary){

        if(firstTime){
            console.log('accept_type:', analytics_summary.accept_type);
            // [TODO] send analytics data to the server ...
        }
    },
    ...
});

**Reply by lparikka** [reply]
Sounds good. Do you think universal enough for people to use?

**Reply by orestbida** [reply]
One other way could be to make use of the `data` field inside the cookie (initially empty/null).

```javascript
onAccept: function(cookie_value, analytics_summary){
    if(!cc.get('data')){
        // [TODO] send analytics data to the server ...
        cc.set('data', { value: true );
    }
}
```

I think it's easy enough to use, perhaphs there are better ways to do this. I'm open to suggestions!

**Reply by OndraM** [reply]
This is what we use right now:

```js
onAccept: (cookie) => {
      const givenLevels = cookieconsent.get('level');
      const acceptedOnlyNecessary = givenLevels.length === 1 && givenLevels[0] === 'necessary'; // we have "necessary" as the default readonly level

    if (acceptedOnlyNecessary) { // aka "reject_all"
        // ...
    }
}
```

However using something like:
```js
onAccept: (cookie, analytics_summary) => {
    if (analytics_summary.accept_type === 'reject_all') {
        // ...
    }
}
```
... would make it a bit cleaner.

**Reply by orestbida** [reply]
Yes, the only "issue" is to log user choices only **once** as @lparikka wanted. 

We'd need to log the preferences of:
- users who have already given consent (and whose preferences are not logged yet)
- new users

The only way to achieve this (as of now) is to rely on the `data` field inside the cookie (by default set to `null`) :
```javascript
if(!cc.get('data')){
    // [TODO] send analytics data to the server ...
    cc.set('data', { value: true );
}
```

Needless to say, the implementation would be different for those who already utilize `data` - which, is not ideal.

**Comment by orestbida** [comment]
@lparikka ,

A solution to this was implemented in v2.7.0. To make it easier — than what was discussed above — a new function prop. called `onFirstAction` was added. It will run only the first time the user makes a choice.

the only parameter of the function - "user_preferences" - returns the following object:

```javascript
{
    accept_type: string,            // 'all', 'necessary', 'custom'
    accepted_categories: string[],  // e.g. ['necessary', 'analytics']
    rejected_categories: string[]   // e.g. ['ads']
}
```

Example:
```javascript
// ...
cc.run({
    // ...
    onFirstAction: function(user_preferences){
        console.log('user accept type:', user_preferences.accept_type);
    },
    // ...
});
```

For more flexibility, a `.getUserPreferences()`method was also added for usage outside the `onFirstAction` function. It returns the same object as "user_preferences" parameter.
